File: algorithms/11_golang/graphs/max_flow_dinic.go
package main

import "fmt"

type Dinic struct {
	adj  [][]int
	to   []int
	cap  []int
	dist []int
	ptr  []int
}

func NewDinic(n int) *Dinic {
	return &Dinic{
		adj:  make([][]int, n),
		dist: make([]int, n),
		ptr:  make([]int, n),
	}
}

func (d *Dinic) AddEdge(u, v, c int) {
	d.adj[u] = append(d.adj[u], len(d.to))
	d.to = append(d.to, v)
	d.cap = append(d.cap, c)
	d.adj[v] = append(d.adj[v], len(d.to))
	d.to = append(d.to, u)
	d.cap = append(d.cap, 0)
}

func (d *Dinic) bfs(s, t int) bool {
	for i := range d.dist {
		d.dist[i] = -1
	}
	d.dist[s] = 0
	q := []int{s}
	for len(q) > 0 {
		u := q[0]
		q = q[1:]
		for _, eid := range d.adj[u] {
			v := d.to[eid]
			if d.dist[v] == -1 && d.cap[eid] > 0 {
				d.dist[v] = d.dist[u] + 1
				q = append(q, v)
			}
		}
	}
	return d.dist[t] != -1
}

func (d *Dinic) dfs(u, t, f int) int {
	if u == t || f == 0 {
		return f
	}
	for ; d.ptr[u] < len(d.adj[u]); d.ptr[u]++ {
		eid := d.adj[u][d.ptr[u]]
		v := d.to[eid]
		if d.dist[v] != d.dist[u]+1 || d.cap[eid] == 0 {
			continue
		}
		pushed := d.dfs(v, t, min(f, d.cap[eid]))
		if pushed > 0 {
			d.cap[eid] -= pushed
			d.cap[eid^1] += pushed
			return pushed
		}
	}
	return 0
}

func (d *Dinic) MaxFlow(s, t int) int {
	flow := 0
	for d.bfs(s, t) {
		for i := range d.ptr {
			d.ptr[i] = 0
		}
		for {
			f := d.dfs(s, t, 1<<60)
			if f == 0 {
				break
			}
			flow += f
		}
	}
	return flow
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	d := NewDinic(6)
	d.AddEdge(0, 1, 16)
	d.AddEdge(0, 2, 13)
	d.AddEdge(1, 2, 10)
	d.AddEdge(1, 3, 12)
	d.AddEdge(2, 1, 4)
	d.AddEdge(2, 4, 14)
	d.AddEdge(3, 2, 9)
	d.AddEdge(3, 5, 20)
	d.AddEdge(4, 3, 7)
	d.AddEdge(4, 5, 4)
	if d.MaxFlow(0, 5) != 23 {
		panic("Dinic: max flow should be 23")
	}
	fmt.Println("[Go Dinic] Max flow verified: 23")
}

File: algorithms/03_cpp/graphs/max_flow_dinic.cpp
/**
 * Dinic's maximum flow algorithm.
 */

#include <vector>
#include <queue>
#include <algorithm>
#include <climits>
#include <cassert>
#include <iostream>

class Dinic {
    struct Edge {
        int to, rev;
        long long cap;
    };
    int n;
    std::vector<std::vector<Edge>> g;
    std::vector<int> level, it;

    bool bfs(int s, int t) {
        std::fill(level.begin(), level.end(), -1);
        std::queue<int> q;
        q.push(s);
        level[s] = 0;
        while (!q.empty()) {
            int u = q.front();
            q.pop();
            for (const auto& e : g[u]) {
                if (e.cap > 0 && level[e.to] == -1) {
                    level[e.to] = level[u] + 1;
                    q.push(e.to);
                }
            }
        }
        return level[t] != -1;
    }

    long long dfs(int u, int t, long long f) {
        if (u == t) return f;
        for (int& i = it[u]; i < static_cast<int>(g[u].size()); ++i) {
            Edge& e = g[u][i];
            if (e.cap > 0 && level[e.to] == level[u] + 1) {
                long long d = dfs(e.to, t, std::min(f, e.cap));
                if (d > 0) {
                    e.cap -= d;
                    g[e.to][e.rev].cap += d;
                    return d;
                }
            }
        }
        return 0;
    }

public:
    explicit Dinic(int n) : n(n), g(n), level(n), it(n) {}

    void addEdge(int u, int v, long long cap) {
        g[u].push_back({v, static_cast<int>(g[v].size()), cap});
        g[v].push_back({u, static_cast<int>(g[u].size()) - 1, 0});
    }

    long long maxFlow(int s, int t) {
        long long flow = 0;
        while (bfs(s, t)) {
            std::fill(it.begin(), it.end(), 0);
            while (long long f = dfs(s, t, LLONG_MAX)) flow += f;
        }
        return flow;
    }
};

int main() {
    Dinic dinic(4);
    dinic.addEdge(0, 1, 16);
    dinic.addEdge(0, 2, 13);
    dinic.addEdge(1, 2, 10);
    dinic.addEdge(1, 3, 12);
    dinic.addEdge(2, 1, 4);
    dinic.addEdge(2, 3, 14);
    assert(dinic.maxFlow(0, 3) == 26);

    Dinic simple(3);
    simple.addEdge(0, 1, 5);
    simple.addEdge(1, 2, 5);
    assert(simple.maxFlow(0, 2) == 5);

    std::cout << "[C++ MaxFlowDinic] Dinic max flow verified." << std::endl;
    return 0;
}
